src/include/txn.i

```diff
@@ -1187,8 +1183,17 @@ __wt_txn_am_oldest(WT_SESSION_IMPL *session)
 
 	WT_ORDERED_READ(session_cnt, conn->session_cnt);
 	for (i = 0, s = txn_global->states; i < session_cnt; i++, s++)
-		if ((id = s->id) != WT_TXN_NONE && WT_TXNID_LT(id, txn->id))
+retry:		if ((id = s->id) != WT_TXN_NONE && WT_TXNID_LT(id, txn->id)) {
+			/*
+			 * If the transaction is still allocating its ID,
+			 * then we spin here until it gets its valid ID.
+			 */
+			if (s->is_allocating != false) {
+				WT_PAUSE();
+				goto retry;
+			}
 			return (false);
+		}
```

Comment:
Can we restructure the code as follows:

```
while ((id = s->id) != WT_TXN_NONE && WT_TXNID_LT(id, txn->id)) {
    if (!s->is_allocating)
        return (false);
    WT_PAUSE();
}
```

Reply:
Actually, most of the change to this function can probably be reverted.  The `__wt_txn_am_oldest` function exists to abort the oldest transaction when activity is blocked (because everything is pinned in cache).  It should be able to ignore any transaction with `is_allocating = true`, though we may want a comment explaining why.  Something like:

```
/*
 * We are checking if this transaction is the oldest one in the system.
 * It is safe to ignore any sessions that are allocating transaction IDs:
 * since we already have an ID, they are guaranteed to be newer.
 */
if (!s->is_allocating && (id = s->id) != WT_TXN_NONE &&
    WT_TXNID_LT(id, txn->id)) {
        ...
```

Reply:
I think we need to consider about simply ignoring the transaction id with `is_allocating = true` because it can miss the real oldest transaction id at that time. A possible history is
1. Thread A (real oldest) set the `is_allocating` flag as `true` and get its transaction id by calling `fetch_and_add(&current, 1)`. And then sleep.
2. Thread B (caller of `__wt_txn_am_oldest`) get its transaction id higher than Thread A, and then call `__wt_txn_am_oldest`. It sees the transaction id of Thread A with `is_allocating = true`, ignore it and decide itself as the oldest one.
3. Thread A begin to execute its transaction with the real oldest id.

To strictly check if this transaction is oldest, I think we need to wait for the transaction id with `is_allocating = true` unless the id value is greater than the current transaction id.
Can you please confirm this @michaelcahill ?

Reply:
@ZombieKim, I think this situation is okay: even if `__wt_txn_am_oldest` returns false in this case, thread B will call it again if eviction is really stuck, and the second time it won't race with any concurrent ID allocation, so it will return `true`.  I think that is preferable to slowing down the common case by waiting for allocating threads.